### ISSUE TITLE
Find Connected Components on images

### ISSUE THREAD

**Issue by jeetkanjani7** [chiggum/cuda-challenge-india#1]
Can this repository be altered to extract blobs and find connected components in images?

**Comment by chiggum** [comment]
Number of components in a binary image can definitely be computed using a part of main.cu. If you can elaborate over your input and expected output, I will try to see if the required procedure can be created by altering my code.

**Comment by jeetkanjani7** [comment]
The input is an RGB image and the output is the number of blobs in the image and an output image with a different colour assigned to each component. I think some functions could be used to generate the desired output (correct me if I am wrong).

**Comment by chiggum** [comment]
Here is one way - Compute a set of unique colors in input image. For each color in the set create an image with the cell with that color as 1 and other colors as 0, in parallel. Now you’ll have a binary image per color in the unique colors set. Use functions in my code to compute number of connected components (n_i) and d_label (d_label_i) for each image in parallel. Finally add all n_i to get total number of connected components and add all d_label_i to get required output image which will have a unique number for each component. Note that the way d_label is computed in the code will ensure unique label for each component. Memory consumption will be of order no. of unique colors times the size of image and time complexity (assuming infinite resources) will approximately be of same order as the CCL part of code (exact time will surely be slightly higher). This approach is based on nested parallelism, thus, will require dynamic parallelism. May be you can start writing a sequential code and I will try to help in between.

**Comment by jeetkanjani7** [comment]
My input is structured as follows: The image(single channel unsigned char) contains regions which have unique intensity values associated with them(background pixels are 0). I need to find connected components with similar intensity value(so two neighbouring non zero pixels with different intensities would lie in different components and I need to assign unique labels or colors to each component). One way as you have mentioned is to have as many number of images as the number of unique intensities and use your functions for finding CC in each followed by reduction, but that would be computationally expensive.  Can you suggest any other approach to solve it efficiently?

**Comment by jeetkanjani7** [comment]
I have added my own custom input [here](https://github.com/jeetkanjani7/cuda-challenge-india-2014/blob/master/custom_input.h). I want the output to be three(the connected 20's and 10's should be treated different components) instead of two. The output '2' makes sense as of now but I am not clear with the output labels. I copied the output labels to the host in 'performCCLnPrintNCC' and they are numbers I can't make sense of, how do I know which ones are part of a connected component? 

This is the output I got with input_h:
 0
 0
 1664644211
 1600218229
 878867811
 1701998435
 1801401137
 1701737061
 1734434668
 1232367205
 1597002574
 1918852920
 1668637797
 1144009061
 1852399986
 1852139329
 1164724596
 810765893
 1969435487
 1917270370
 1968268393
 1231385957
 1816479081
 1417037125
 827612976
Sorry for the lame question.
